staging/src/k8s.io/client-go/tools/leaderelection/resourcelock/configmapleaselock.go

```diff
@@ -0,0 +1,82 @@
+/*
+Copyright 2019 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package resourcelock
+
+import (
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+)
+
+// ConfigMapLeaseLock is used for migration between the configmap and the lease
+type ConfigMapLeaseLock struct {
+	ConfigMapLock *ConfigMapLock
+	LeaseLock     *LeaseLock
+}
+
+// Get returns the older election record of the configmap and the lease
+func (cmll *ConfigMapLeaseLock) Get() (*LeaderElectionRecord, error) {
+	configMapLock, err := cmll.ConfigMapLock.Get()
+	if err != nil {
+		return nil, err
+	}
+	leaseLock, err := cmll.LeaseLock.Get()
+	if err != nil {
+		return nil, err
+	}
+	// check if the older resource lock is released
+	if leaseLock.RenewTime.Before(&configMapLock.RenewTime) {
+		return leaseLock, nil
+	}
+	return configMapLock, nil
+}
+
+// Create attempts to create both configmap and lease
+func (cmll *ConfigMapLeaseLock) Create(ler LeaderElectionRecord) error {
+	err := cmll.ConfigMapLock.Create(ler)
+	if err != nil && !apierrors.IsAlreadyExists(err) {
```

Comment:
if this already exists, and we do not verify the object that already exists matches the LeaderElectionRecord we intended to create, doesn't that cause problems?

Reply:
doesn't this also put the configmap holder in an invalid state (it makes the internal `cm` variable nil in an AlreadyExists case)?

Reply:
Not sure I understand it - this could potentially be a problem already (even without multilock).

Reply:
I think there is a assumption here:
We will call Get() firstly, if it is not found. We would not meet already exists error here. And we can initial cm variable successfully.
If it already exist, function Get() will initial cm variable.

And as wojtek said, this could potentially be a problem already (even without multilock).

Reply:
> this could potentially be a problem already (even without multilock).

I don't think so... we don't mask AlreadyExists errors in ConfigMapLock#Create, so tryAcquireOrRenew fails and retries, where the next Get() would return the object

Reply:
> I don't think so... we don't mask AlreadyExists errors in ConfigMapLock#Create, so tryAcquireOrRenew fails and retries, where the next Get() would return the object

I think I get your point.
If two clients start for a initial cluster, they may Get() and both get NotFound error, then try to Create() and both return true.
The problem is Get() and Create() is not atomic.

How about masking AlreadyExists errors when the primary lock is created by the client.
The reason about masking AlreadyExists errors is that Get() return NotFound error, but we don't know which lock is not found, so we try to create both of them.
It the primary lock is belong to the client, the secondary lock is NotFound, and we would create it, else if the primary is created by other client, we would return error and tryAcquireOrRenew would fail.